File: workers/matrix.rb
stations = {}

ARGV.each do |filename|
   STDERR.puts filename
   file = File.open("#{filename}")
   samples = {}
   file.readlines.each do |line|
      s=line.split(",")
      sample = s[0]
      freq   = s[1]
      samples[sample] = [] if samples[sample].nil?
      samples[sample] << freq.to_i
   end
   STDERR.puts "Done reading #{filename}"
   
   file.close
   # sum up the the frequency for each sample at this station
   samples.each do |sample,freq|
      totalFreq = freq.reduce(:+)
      samples[sample] = totalFreq
   end
   STDERR.puts "Done summing frequencies"
   stations[filename] = samples
end
allSamples = []
# get a list of all the sample names
stations.invert.keys.each{|samples| allSamples << samples.keys}
printf "Sample\t"
stations.keys.each{|station| printf "%10s\t", station}
puts
allSamples.flatten.uniq.sort.each do |sample|
   printf "#{sample}\t"
   stations.each do |station,samp|
      padding = station.length
      if ! samp[sample].nil?
         printf("%10d\t", samp[sample])
      else
         printf("%10d\t", 0)
      end
   end
   puts
end

File: workers/matrix.pl
use strict;
use warnings;

my %stations;
my %samples;
my @allSamples;
my $freq;
my $sample;
my $file;
my @stations;
my $filename;
for(my $i=0; $i < $#ARGV+1; $i++){
   %samples = ();
   $filename = $ARGV[$i];
   open($file, "<", $filename)
           or die "cannot open < $ARGV[$i]: $!";
   print STDERR "$filename\n";
   while ( ! eof($file) ) {
      my $line = readline($file);
      chomp($line);
      my @s = split(',', $line);
      $sample = $s[0];
      $freq = $s[1];
      push @allSamples, $sample;
      push @{$samples{$sample}}, $freq;
   }
   print STDERR "Done reading $filename\n";
   close $file;

   foreach my $key (keys %samples){
      my @freqs = @{$samples{$key}};
      my $totalFreqs = '0';
      map {$totalFreqs += $_} @freqs;
      $samples{$key} = $totalFreqs;
   }
   print STDERR "Done summing frequencies\n";
   %{$stations{$filename}} = %samples;
}
my @uniqSamples = sort (uniq(@allSamples));

my $output = "";
$output .= sprintf "Sample\t";
foreach my $key (keys %stations){ $output .= sprintf("%10s\t",$key); }
$output .= sprintf "\n";
foreach my $sample (@uniqSamples) {
   $output .= sprintf "$sample\t";
   foreach my $station (keys %stations){
      $freq = $stations{$station}{$sample};
      if (exists $stations{$station}{$sample}) {
         $output .= sprintf("%10d\t", $freq);
      } else {
         $output .= sprintf ("%10d\t", 0);
      }
   }
   $output .= sprintf "\n";
}
print $output;
#use Data::Dumper;
#print Dumper \@uniqSamples;


sub uniq {
    my %seen;
    grep !$seen{$_}++, @_;
}
